Topics: update CTV to cover 4+ years of discussion (#1816)
bitcoinops/bitcoinops.github.io: _topics/en/op_checktemplateverify.md

excerpt: >
  **OP_CHECKTEMPLATEVERIFY (CTV)** is a proposed new opcode that takes a
  commitment hash as a parameter and requires any transaction executing

>   the opcode to include a set of outputs that match the commitment.  This

Reviewer: probably better:

constrains the transaction to have an exact set of

nVersion
nLockTime
scriptSig hash (if any are non-null, otherwise restricted to all be null)
input count
sequences hash
output count
outputs hash
input index

  - title: Covenants
    link: topic covenants
---
CTV has been suggested for use with a variety of protocols:

- **Congestion control transactions:** a spender pays a single output
  that uses CTV to commit to a follow-up transaction that pays multiple
  outputs.  When the initial transaction is confirmed to a suitable
  depth, this can assure the parties with outputs in the follow-up
  transaction that they will eventually be paid (e.g. when feerates are
  lower).  This two-step process can probably be used anywhere [payment
  batching][topic payment batching] is an option, although it requires
  extra communication to receivers for them to learn that they have been

>   paid prior the follow-up transaction being confirmed.

Reviewer 2: ```suggestion
  paid prior to the follow-up transaction being confirmed.
```

  all of them in agreement to spend a portion of the funds.  This is
  possible today by having each participant sign the unilateral exit
  transaction for the next state before they sign the next state, but
  this requires two rounds of signing.  When the number of parties
  needed to sign is large, each round can take a long time.  CTV reduces
  the amount of signing to one round.

  Additionally, when one party wants to unilaterally exit a joinpool or
  channel factory, some other parties may want to remain in the joinpool
  (or form subsets of the original joinpool).  This can minimize the
  amount of onchain fees that need to be paid by each party.  Remaining
  in an existing joinpool requires each party agree to the outcomes for
  each other party exiting.  Using presigned exits, the most efficient
  onchain version of this requires generating and transmitting O(n!)
  offchain signatures for _n_ parties, making it impractical for more
  than a few users to participate in any individual joinpool with
  per-party exit paths.  Using CTV, each signature operation can be
  replaced with a CTV operation, which leaves the computational

>   complexity but doesn't require transmitting any additional data---each

Author: I'm not 100% sure this description is accurate.  I need to do some additional research or rewrite it to be less specific before we publish.

Reviewer: probably makes sense to discuss ARK and ARK v2 here, which AFAIU both work under CTV as a standalone upgrade?

Author: Made text vague about the actual complexity (to address my point) and added back link from our Ark topic page to this CTV to partially address @JeremyRubin's point.  I forgotten most about what I learned about Ark last year and haven't learned anything about Ark v2 yet, and I don't want to block updating this page on learning about them to maybe add a sentence or two.

  user can derive the necessary data on their own (which they would need

>   to do any way to validate the script they're paying).  Again, only the

Reviewer 2: ```suggestion
  to do anyway to validate the script they're paying).  Again, only the
```

  one round of signatures for each state transition would need to be

>   communicated.  Alternative consensus change proposal specifically

Reviewer 2: ```suggestion
  communicated.  Alternative consensus change proposals specifically
```

  contract---minimizing bandwidth and storage.  A few signatures and
  adaptors are still needed, but the vast majority are eliminated.

## Alternatives to CTV

Many alternatives to CTV have been proposed.  The following have been
covered by Optech:

- **Presigned transactions:** anything CTV can do can also be done using
  presigned transactions and secure deletion of ephemeral keys.
  However, this requires perfectly reliable backups and deletion.  For
  CTV, deletion is not required and the amount of data that needs to be
  reliably backed up consists of only the data a user expects to put
  onchain anyway (some of which it may be possible to compute as needed).

- **SIGHASH_ANYPREVOUT:** this [proposal][topic sighash_anyprevout],
  which was designed to [improve payment channels][topic eltoo] before
  the announcement of CTV, can substitute for CTV in many cases (but not
  all), although it requires extra onchain vbytes to accomplish patterns
  that CTV is optimized for.  At least one [variant of
  SIGHASH_ANYPREVOUT][poinsot apo] has been proposed to extend the
  number of CTV usecases covered, although it still requires extra

>   vbytes.

Reviewer: The state of the art of LN-Symmetry has shown that CTV + CSFS + IKey is more bytewise efficient than APO for payment channels, although CTV + CSFS + IKey + APO is slightly more efficient.

cc @reardencode

Author: Leaving as-is.  This section isn't discussing the efficiency of APO vs CTV for payment channels, it's discussing APO as a generic alternative to CTV (which it isn't, not 100%).

- **OP_CAT + OP_CHECKSIGFROMSTACK:** the composition of the proposed
  [OP_CAT][topic op_cat] and [OP_CHECKSIGFROMSTACK][topic
  op_checksigfromstack] opcodes can emulate CTV, although this may
  require significantly more vbytes.

- **OP_TX and OP_TXHASH:** are [roughly][op_tx] equivalent
  [proposals][op_txhash] to enable CTV features in a way that might be

>   easier to extend in the future.  Depending on the exact design, they

Reviewer: CTV is actually extensible via soft fork, OP_TX and OP_TXHASH are basically a combination of tweaks to CTV's commitment structure + a set of sighash flags precomitting to all possible uses of CTV, rather than CTV's choice to leave those up for future extensibility.

Author: Leaving as-is.  I think the "might" in "might be easier to extend in the future" is open-ended enough here.

  may require slightly more vbytes to use than CTV.

- **Exploding keys:** this [proposal][dryja exploding keys] provides the
  core feature of CTV: the ability for a transaction output to commit to
  the outputs of the transaction that will spend it.  CTV accomplishes
  this using a hash digest on the stack; exploding keys accomplishes it
  using a [pay-to-contract][topic p2c]-like commitment to the hash
  digest.  Exploding keys can, in theory, save a few vbytes compared to

>   CTV---although it also loses some of CTV's flexibility.

Reviewer: exloding keys is as efficient as raw CTV -- raw bare CTV is just 32 bytes + 1 opcode + 1 pushdata.

https://x.com/JeremyRubin/status/1782974382391935090 see discussion here.

Reviewer: main advantage is creating a new anonymity set for raw ctv.

computational costs are also a negative for things like DLCs, since all the curve ops are slow.

  but without enabling [recursive covenants][topic covenants].  However,
  [some critics][corallo flexible] would [prefer][corallo costs] to see
  a more general covenant solution.  They worry that many of the people
  who ask for CTV now will not be satisfied with it.  They may later ask
  for an improved construction that could render CTV unnecessary.  They
  advocate instead for effort to be devoted to finding a solution that
  will fully satisfy user desires.

- **Not generic enough:** CTV is almost maximally efficient at allowing
  an output to require it be spent to certain future outputs.  It
  additionally provides some flexibility in the construction of that
  spending transaction, mainly related to which inputs it includes.
  However, its capabilities overlap to a significant degree with other
  [covenant][topic covenants] proposals, so it has been
  [suggested][oconnor composable] that it would be better to use
  composable elements to replicate the features of CTV.  Those elements
  may also be useful on their own, or they may be extensible in ways
  that don't necessarily make sense for CTV.  This is the origin of
  proposals such as `OP_TX` (described previously).

> - **Makes script harder to analyze:** CTV tracks information about the

Reviewer 2: Maybe 

```suggestion
- **Makes scripts harder to analyze:** CTV tracks information about the
```

or 

```suggestion
- **Makes Script harder to analyze:** CTV tracks information about the
```

  executed opcode that precedes it as part of its protection against
  recursive covenants.  This is abnormal among opcodes
  (`OP_CODESEPARATOR` in its interaction with `OP_CHECKSIG` and variants
  being the only current opcode we're aware of that requires similar
  tracking).  At least one reviewer argues that dealing with the
  abnormal state requirements of CTV [adds][oconnor complexity] "much

>   more complexity to anyone trying to formally model Script semantics".

Reviewer: this was true for OP_COSHV and removed by the CTV proposal in response to the feedback. Just a normal script semantic opcode now. @roconnor-blockstream

- **CTV and OP_VAULT:** the [BIP345][] `OP_VAULT` proposed soft fork is
  designed (in its latest versions) to work directly with CTV for
  committing to spending operations for creating dynamic [vaults][topic
  vaults].

- **OP_VAULT alternative using MATT and CTV:** the [MATT][topic acc]
  proposed soft fork has been described as being able to almost emulate
  `OP_VAULT` when combined with CTV and two other opcodes.  This proposal
  appears mainly designed to showcase the flexibility of MATT rather
  than suggest activating MATT and CTV instead of `OP_VAULT` and CTV.

- **CTV and OP_CHECKSIGFROMSTACK:** the [OP_CHECKSIGFROMSTACK][topic
  op_checksigfromstack] (CSFS) proposed soft fork allows verifying that
  a signature on the stack commits to another item on the stack.  This
  is in comparison to existing signature verification operations in
  Bitcoin that verify that a signature commits to a large amount of data
  and metadata for the executing transaction.  It's [proposed][news285
  lnhance] that CTV and CSFS together can emulate the proposed
  `SIGHASH_ANYPREVOUT` (APO) soft fork and so can enable versions of
  [LN-Symmetry][topic eltoo] (eltoo) and [Ark][topic ark]-style

>   [joinpools][topic joinpools].

Reviewer: afaiu ark does not require CSFS?

## History

CTV was originally proposed as `OP_CHECKOUTPUTSHASHVERIFY` (COSHV) and
was briefly renamed to `OP_SECURETHEBAG` before becoming

> `OP_CHECKTEMPLATEVERIFY`.  The earliest public version of the proposal

Reviewer: more precisely, each proposal has been "unchanged". they are 3 distinct opcode proposals, but COSHV and STB are withdrawn in favour of CTV. So CTV is the draft proposal that was formalized, implemented, and submitted to the BIP repo. the others were WIP.

This is a bit different than e.g. Taproot, where many changes were made to what taproot was over the years before final inclusion. CTV is unchanged.

allowed use of inputs differently than the current proposal.  A later

> [update][ctv extensions] to the proposal allowed use of 20-byte hash

Reviewer: It's a proposed extension using CTV's upgrade mechanism -- I think one that shouldn't be taken personally v.s. a better CSFS -- for solving some issues in LN Symmetry people were trying to figure out how to do with only LNHance and default CSFS.

Nuance is that CTV's BIP is unchanged by an extension.
